### PR TITLE
Issue #8025. Ignoring HEADER and DATA frames for streams that may have existed in the past.

### LINKED ISSUE #8025: HTTP/2 - Race Condition with RST_STREAM
### Expected behavior
When a RST_STREAM is sent to the remote, any future incoming DATA frames should be ignored.

### Actual behavior
Sometimes an error is thrown instead. (See https://github.com/grpc/grpc-java/issues/3548)

### Steps to reproduce
Send a RST_STREAM frame to a remote endpoint.

### Netty version
4.1.23.Final

### JVM version (e.g. `java -version`)
Java 8

### OS version (e.g. `uname -a`)
Any (observed on Mac and Linux).

For a bit more context. Sending a RST_STREAM will eventually trigger this code in Http2ConnectionHandler, https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L589 . This removes the stream from the active streams metadata and any subsequent DATA frames that come in will be considered as an unknown stream. I was thinking we could just check if the stream may have existed and if so just ignore the DATA frame altogether. So instead of [this](https://github.com/netty/netty/blob/501662a77f8166382eefd18a8620983d27f59d4a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L535) do something like this

```java
verifyStreamMayHaveExisted(streamId);
return true;
```

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Correctly ignore frames that come in after a RST_STREAM is sent.

Modification:

Ignoring frames that are for a stream that may have existed in the past instead of throwing an exception.

Result:

Fixes #8025.

### CODE REVIEW

**Comment by normanmaurer** [comment]
This is https://github.com/netty/netty/pull/8028 with followup cleanup... Sorry had to close the original PR to make progress :/

**Comment by normanmaurer** [comment]
I would love to finally pull this in...

**Comment by bryce-anderson** [comment]
iirc, this work was spawned by https://github.com/grpc/grpc-java/issues/3548 which is basically some log entries that looks spooky but in reality are expected in the case of stream resets. It's been a while, but IMO, why can't we just decrease the log message to debug? I don't see the harm in sending additional RST frames since this isn't expected to be that expensive or common.

**Comment by zrss** [comment]
cc

[Alluxio worker keep output this stream error warning](https://github.com/Alluxio/alluxio/issues/9637)

**Comment by normanmaurer** [comment]
@ejona86 WDYT ?

**Comment by normanmaurer** [comment]
@ejona86 any thoughts ?

**Comment by ejona86** [comment]
Reducing the log level sounds fine to me. But this is also like the 4-5th time we've "fixed" the logic surrounding this case, which means I'm happy if someone is able to spend the time to think seriously about it.

But yeah, I would just reduce the log level, especially given how the log messages I've seen have been triggered.

**Comment by bryce-anderson** [comment]
We see these being logged by Netty, almost certainly in `Http2FrameCodec.onHttp2UnknownStreamError`, and they 'are definitely spooky to see for our HTTP/2 users.

**Comment by normanmaurer** [comment]
@bryce-anderson may want to do a PR to "fix" this ?

**Comment by bryce-anderson** [comment]
I can take a look, though it may be a few days. I don't recall what this PR does, but from the comments it seems like a reasonable approach as well.

**Comment by normanmaurer** [comment]
@bryce-anderson that's fine :)

**Comment by normanmaurer** [comment]
Even with log level adjusted this PR may still make sense to pull in... @ejona86 @bryce-anderson @mosesn @carl-mastrangelo @trustin WDYT ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamRingBuffer.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        return elements.length;
+    }
+
+    /**
+     * Add the given {@code streamId}
+     * @param streamId the ID
+     * @return {@code true} if could be added, false if it was not added as this ring-buffer already contained it.
+     */
+    boolean add(int streamId) {
+        assert streamId >= 0;
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamRingBuffer.java]: Maybe this should be `> 0`? It seems unlikely we'll close the connection stream and still worry about this.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamRingBuffer.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    private final int[] elements;
+    private int oldestIdx;
+
+    Http2StreamRingBuffer() {
+        this(32);
+    }
+
+    Http2StreamRingBuffer(int capacity) {
+        elements = new int[MathUtil.findNextPositivePowerOfTwo(capacity)];
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamRingBuffer.java]: Does it matter if we use a power of 2? It doesn't look like we do any modulo or division operations.

**Comment by normanmaurer** [comment]
@bryce-anderson any updates related to a patch for the logging ?

**Comment by bryce-anderson** [comment]
Not yet, hoping to get to it this week. Sorry for the delay. With regard to
this change, it looks sound although I'm on the fence as to whether it's
worth the extra code. Perhaps someone else has a strong opinion one way or
the other.

**Comment by bryce-anderson** [comment]
@normanmaurer, I was thinking about this patch over the weekend and something came to mind that is all but certainly not worth the effort. 😄 What if we kept the streams that have been reset around in a terminal CLOSED state for a duration. That would keep a marker around for a bit and after a reasonable time frame (and 'reasonable' is probably the sticking point of this idea) we evict them from the `Http2Connection`.

**Comment by normanmaurer** [comment]
lets just close this for now

**Comment by PSNRigner** [comment]
Was a fix deployed for this ? At least for the logger ? Because this is a real issue for us.

**Comment by normanmaurer** [comment]
Yes and it will be Part of the next release

**Comment by bhavsi** [comment]
When is the release or has it already happened? I am circling around in a loop between links on this thread and links on the page it navigates to, ending back here.

**Comment by normanmaurer** [comment]
@bhavsi yes it was released as part of 4.1.44.Final as stated in the original issue

**Pull request closed without merging**